Mathlib/Dynamics/SymbolicDynamics/Basic.lean

```diff
@@ -0,0 +1,745 @@
+
+Intuitively, a pattern is a "partial configuration" (like a finite word),
+specifying finitely many values of a configuration in `G → A`. Patterns are
+the basic building blocks used to define subshifts via forbidden configurations.
+Note that each pattern corresponds to a cylinder, which is the set of configurations
+which agree with this pattern on its support. -/
+structure Pattern (A : Type*) (G : Type*) where
+  /-- Finite support of the pattern. -/
+  support : Finset G
+  /-- The value (symbol) at each point of the support. -/
+  data : support → A
+
+section Dominos
+
+variable (G : Type*) [Monoid G]
+
+/-- A *domino* is a pattern supported on exactly two positions (sometimes
+also called cells) `{i, j}`, specifying the value `ai` at `i` and the value `aj` at `j`.
+
+In symbolic dynamics, dominos (two-cell patterns) are the simplest nontrivial
+building blocks: they describe local adjacency conditions between two sites
+of a configuration. -/
+def domino {A : Type*} (i j : G) (ai aj : A) : Pattern A G := by
+  classical
+  exact
+    { support := ({i, j} : Finset G)
+      data := fun ⟨z, _hz⟩ => if z = i then ai else aj }
+
+end Dominos
+
+section Forbids
+
+variable {A G : Type*} [Monoid G]
+
+/-- `p.mulOccursInAt x g` means that the finite pattern `p` appears in the configuration `x`
+at position `g`.
+
+Formally: for every position `h` in the support of `p`, the value of the configuration
+at `h * g` coincides with the value specified by `p` at `h`.
+
+Intuitively, if you shift the configuration `x` by `g` (using `mulShift g`),
+then on the support of `p` you exactly recover the pattern `p`. This is the basic
+notion of "pattern occurrence" used to define subshifts via forbidden patterns. -/
+@[to_additive Pattern.occursInAt
+/-- `p.occursInAt x g` means that the finite pattern `p` appears in the configuration `x`
+at position `g`.
+
+Formally: for every position `h` in the support of `p`, the value of the configuration
+at `h + g` coincides with the value specified by `p` at `h`.
+
+Intuitively, if you shift the configuration `x` by `g` (using `shift g`),
+then on the support of `p` you exactly recover the pattern `p`. This is the basic
+notion of "pattern occurrence" used to define subshifts via forbidden patterns. -/]
+def Pattern.mulOccursInAt (p : Pattern A G) (x : G → A) (g : G) : Prop :=
+  ∀ (h) (hh : h ∈ p.support), x (h * g) = p.data ⟨h, hh⟩
+
+/-- `mulForbids F` is the set of configurations that avoid every pattern in `F`.
+
+Formally: `x ∈ mulForbids F` if and only if for every pattern `p ∈ F` and every
+group element `g : G`, the pattern `p` does not occur in `x` at position `g`.
+
+Intuitively, `mulForbids F` is the shift space defined by declaring the finite set
+(or family) of patterns `F` to be *forbidden*. A configuration belongs to the subshift if and only
+it avoids all the forbidden patterns. -/
+@[to_additive forbids
+/-- `forbids F` is the set of configurations that avoid every pattern in `F`.
```

Comment:
`forbids` of `forbidden`? `F` is forbidden, but I don't really see who or what forbids, so I'd go for the latter.